pkg/blocks/basicstep.go

```diff
@@ -31,108 +31,35 @@ import (
 	"github.com/facebookincubator/ttpforge/pkg/logging"
 	"github.com/facebookincubator/ttpforge/pkg/outputs"
 	"go.uber.org/zap"
-	"gopkg.in/yaml.v3"
+)
+
+const (
+	ExecutorPython     = "python3"
+	ExecutorBash       = "bash"
+	ExecutorSh         = "sh"
+	ExecutorPowershell = "powershell"
+	ExecutorRuby       = "ruby"
+	ExecutorBinary     = "binary"
+	ExecutorCmd        = "cmd.exe"
 )
 
 // BasicStep is a type that represents a basic execution step.
 type BasicStep struct {
-	*Act        `yaml:",inline"`
-	Executor    string     `yaml:"executor,omitempty"`
-	Inline      string     `yaml:"inline,flow"`
-	Args        []string   `yaml:"args,omitempty,flow"`
-	CleanupStep CleanupAct `yaml:"cleanup,omitempty"`
+	actionDefaults `yaml:"-"`
+	Executor       string                  `yaml:"executor,omitempty"`
+	Inline         string                  `yaml:"inline,flow"`
+	Environment    map[string]string       `yaml:"env,omitempty"`
+	Outputs        map[string]outputs.Spec `yaml:"outputs,omitempty"`
 }
 
 // NewBasicStep creates a new BasicStep instance with an initialized Act struct.
 func NewBasicStep() *BasicStep {
-	return &BasicStep{
-		Act: &Act{
-			Type: StepBasic,
-		},
-	}
-}
-
-// UnmarshalYAML custom unmarshaler for BasicStep to handle decoding from YAML.
-func (b *BasicStep) UnmarshalYAML(node *yaml.Node) error {
-	type BasicStepTmpl struct {
-		Act         `yaml:",inline"`
-		Executor    string    `yaml:"executor,omitempty"`
-		Inline      string    `yaml:"inline,flow"`
-		Args        []string  `yaml:"args,omitempty,flow"`
-		CleanupStep yaml.Node `yaml:"cleanup,omitempty"`
-	}
-
-	var tmpl BasicStepTmpl
-	// there is an issue with strict fields not being managed https://github.com/go-yaml/yaml/issues/460
-	if err := node.Decode(&tmpl); err != nil {
-		return err
-	}
-
-	b.Act = &tmpl.Act
-	b.Args = tmpl.Args
-	b.Executor = tmpl.Executor
-	b.Inline = tmpl.Inline
-
-	if b.IsNil() {
-		return b.ExplainInvalid()
-	}
-
-	// we do it piecemiel to build our struct
-	if tmpl.CleanupStep.IsZero() || b.Type == StepCleanup {
-		return nil
-	}
-
-	logging.L().Debugw("step", "name", tmpl.Name)
-	cleanup, err := b.MakeCleanupStep(&tmpl.CleanupStep)
-	logging.L().Debugw("step", "err", err)
-	if err != nil {
-		return err
-	}
-
-	b.CleanupStep = cleanup
-
-	return nil
-}
-
-// Cleanup is an implementation of the CleanupAct interface's Cleanup method.
-func (b *BasicStep) Cleanup(execCtx TTPExecutionContext) (*ActResult, error) {
```

Comment:
This refactor removes the need for defining most of these methods, as the `Action` (formerly `Step`) interface has been streamlined.